File: LeetCodePractice/src/DP/TargetSum.java
package DP;
/*You are given a list of non-negative integers, a1, a2, ..., an, and a target, S. 
 * Now you have 2 symbols + and -. 
 * For each integer, you should choose one from + and - as its new symbol.

Find out how many ways to assign symbols to make sum of integers equal to target S.
Example
Input: nums is [1, 1, 1, 1, 1], S is 3. 
Output: 5
Explanation: 

-1+1+1+1+1 = 3
+1-1+1+1+1 = 3
+1+1-1+1+1 = 3
+1+1+1-1+1 = 3
+1+1+1+1-1 = 3

There are 5 ways to assign symbols to make the sum of nums be target 3.*/
//Let P denotes a set of nums have + sign before it
//Let N denotes a set of nums have - sign before it
//P U N = nums 
//sum(P) - sum(N) = S            
//sum(P) - sum(N) + sum(P) + sum(N) = target + sum(P) + sum(N)
//2 * sum(P) = S + sum(nums);
//sum(P) = (S + sum(nums)) / 2; ~ 0:1 backpack problem         https://www.youtube.com/watch?v=zks6mN06xdQ
//new target is (S + sum(nums)) / 2
public class TargetSum {
	/**
     * @param nums: the given array
     * @param s: the given target
     * @return: the number of ways to assign symbols to make sum of integers equal to target S
     */
	public int findTargetSumWays(int[] nums, int S) {
        int totalSum = 0;
        int len = nums.length;
        for(int i = 0; i < len; i++){
            totalSum += nums[i];
        }
        if(totalSum > 1000 || S > 1000){
            return 0;
        }
        if(len > 20){
            return 0;
        }
        if((S + totalSum) % 2 != 0){
            return 0;
        }
        int target = (S + totalSum) / 2;
        //mem[i][j] means number of subset from nums[0: j - 1] has sum equals to i
        int[][] dp = new int[target + 1][len + 1];
        dp[0][0] = 1;
        for(int i = 0; i <= target; i++){
            for(int j = 1; j <= len; j++){
            	dp[i][j] = dp[i][j - 1];
                if(i >= nums[j - 1]){
                    dp[i][j] += dp[i - nums[j - 1]][j - 1];
                }
            }
        }
        return dp[target][len];     
    }
}

File: LeetCodePython/TargetSum.py
"""
You are given an integer array nums and an integer target.

You want to build an expression out of nums by adding one of the symbols '+' and '-' before each integer in nums and then concatenate all the integers.

For example, if nums = [2, 1], you can add a '+' before 2 and a '-' before 1 and concatenate them to build the expression "+2-1".
Return the number of different expressions that you can build, which evaluates to target.

Example 1:
Input: nums = [1,1,1,1,1], target = 3
Output: 5
Explanation: There are 5 ways to assign symbols to make the sum of nums be target 3.
-1 + 1 + 1 + 1 + 1 = 3
+1 - 1 + 1 + 1 + 1 = 3
+1 + 1 - 1 + 1 + 1 = 3
+1 + 1 + 1 - 1 + 1 = 3
+1 + 1 + 1 + 1 - 1 = 3

Example 2:
Input: nums = [1], target = 1
Output: 1

Constraints:
1 <= nums.length <= 20
0 <= nums[i] <= 1000
0 <= sum(nums[i]) <= 1000
-1000 <= target <= 1000
"""
from typing import List


class TargetSum:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        target = target + sum(nums)
        if target < 0 or target % 2 != 0:
            return 0
        target = target // 2
        nums_size = len(nums)
        dp = [[0] * (nums_size + 1) for _ in range(target + 1)]
        dp[0][0] = 1
        for i in range(target + 1):
            for j in range(1, nums_size + 1):
                dp[i][j] = dp[i][j - 1]
                if i >= nums[j - 1]:
                    dp[i][j] += dp[i - nums[j - 1]][j - 1]
        return dp[target][nums_size]
